Updated version of the boosted double SV tagger

ferencek opened the pull request:
Updated version of the boosted double SV tagger which is also added to slimmed ak8 jets in MiniAOD.

ferencek committed:
updated version of the double-b tagger

ferencek committed:
updated track counting

ferencek committed:
updated track selection DeltaR cut

ferencek committed:
fixed incorrect variable type, removed redundant switch cases

ferencek committed:
further updates to track counting

ferencek committed:
added double-b tagger to slimmed ak8 jets in MiniAOD

cmsbuild commented:
A new Pull Request was created by @ferencek (Dinko Ferencek) for CMSSW_8_0_X.

It involves the following packages:

PhysicsTools/PatAlgos
RecoBTag/SecondaryVertex
RecoBTag/SoftLepton
RecoBTau/JetTagComputer


@cvuosalo, @monttj, @cmsbuild, @slava77, @vadler, @davidlange6 can you please review it and eventually sign? Thanks.
@rappoccio, @imarches, @ahinzmann, @acaudron, @mmarionncern, @jdolen, @nhanvtran, @schoef, @ferencek, @gpetruc, @mariadalfonso, @pvmulder, @TaiSakuma this is something you requested to watch as well.
@degano you are the release manager for this.

Following commands in first line of a comment are recognized

* **+1|approve[d]|sign[ed]**: L1/L2's to approve it
* **-1|reject[ed]**: L1/L2's to reject it
* **assign &lt;category&gt;[,&lt;category&gt;[,...]]**: L1/L2's to request signatures from other categories
* **unassign &lt;category&gt;[,&lt;category&gt;[,...]]**: L1/L2's to remove signatures from other categories
* **hold**: L1/all L2's/release manager to mark it as on hold
* **unhold**: L1/user who put this PR on hold
* **merge**: L1/release managers to merge this request
* **[@cmsbuild,] please test**: L1/L2 and selected users to start jenkins tests
* **[@cmsbuild,] please test with cms-sw/cmsdist#&lt;PR&gt;**: L1/L2 and selected users to start jenkins tests using externals from cmsdist

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-12442/9758/summary.html

cvuosalo commented on RecoBTag/SecondaryVertex/src/CandidateBoostedDoubleSecondaryVertexComputer.cc:
```diff
@@ -13,39 +20,73 @@ CandidateBoostedDoubleSecondaryVertexComputer::CandidateBoostedDoubleSecondaryVe
+  float tau2_trackEtaRel_0 = -1., tau2_trackEtaRel_1 = -1., tau2_trackEtaRel_2 = -1.;
+  float tau1_vertexMass = -1., tau1_vertexEnergyRatio = -1., tau1_vertexDeltaR = -1., tau1_flightDistance2dSig = -1.;
+  float tau2_vertexMass = -1., tau2_vertexEnergyRatio = -1., tau2_vertexDeltaR = -1., tau2_flightDistance2dSig = -1.;
+  float jetNTracks = 0, nSV = 0, tau1_nSecondaryVertices = 0, tau2_nSecondaryVertices = 0;
```
Please use static const float for these magic numbers like -50. and -19.

ferencek replied:
These are dummy default values. The variables get updated later so they cannot be declared const.

cvuosalo replied:
@ferencek:
I was not suggesting that variables like trackSip3dSig_3 be declared const.  What's needed is to add ` static constexpr float dummySignificance = -50.0f;` and then use it in place of -50 throughout the file.

ferencek replied:
@cvuosalo, thanks for the clarification. I will try to implement this by Wednesday.

cvuosalo commented on RecoBTag/SecondaryVertex/src/CandidateBoostedDoubleSecondaryVertexComputer.cc:
```diff
@@ -54,111 +95,475 @@ float CandidateBoostedDoubleSecondaryVertexComputer::discriminator(const TagInfo
+        direction = GlobalVector(currentAxes[0].px(), currentAxes[0].py(), currentAxes[0].pz());
+    }
+    else if (currentAxes.size() > 0)
+      direction = GlobalVector(currentAxes[0].px(), currentAxes[0].py(), currentAxes[0].pz());
```
Much simpler and to avoid code duplication:
```
int index = 0;
if (currentAxes.size() > 1 && reco::deltaR2(ptrack,currentAxes[1]) < reco::deltaR2(ptrack,currentAxes[0]))
    index = 1;
direction = GlobalVector(currentAxes[index].px(), currentAxes[index].py(), currentAxes[index].pz());
```

ferencek replied:
Done.

cvuosalo commented on RecoBTag/SecondaryVertex/src/CandidateBoostedDoubleSecondaryVertexComputer.cc:
```diff
@@ -54,111 +95,475 @@ float CandidateBoostedDoubleSecondaryVertexComputer::discriminator(const TagInfo
+
+    temp = tau1_vertexDeltaR;
+    tau1_vertexDeltaR= tau2_vertexDeltaR;
+    tau2_vertexDeltaR= temp;
```
This line is a dead assignment flagged by the static analyzer. Please delete it.

ferencek committed:
implemented Carl's suggestions

cvuosalo commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
Pull request #12442 was updated. @cvuosalo, @monttj, @cmsbuild, @slava77, @vadler, @davidlange6 can you please check and sign again.

cmsbuild commented:
+1
Tested at: ff048c46f604f638ab91531481116a5986af7a85
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-12442/9874/summary.html

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-12442/9874/summary.html

ferencek committed:
turning more magic numbers into static consts

ferencek commented:
@cvuosalo, with a bit of delay, I think I now implemented all of your suggestions.

cmsbuild commented:
Pull request #12442 was updated. @cvuosalo, @monttj, @cmsbuild, @slava77, @vadler, @davidlange6 can you please check and sign again.

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 286dc62c7bbdbc095f2317cbf74b7f5e52d45658
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-12442/9973/summary.html

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-12442/9973/summary.html

cvuosalo commented:
Extended tests in progress...

cvuosalo commented on RecoBTag/SecondaryVertex/src/CandidateBoostedDoubleSecondaryVertexComputer.cc:
```diff
@@ -54,111 +95,474 @@ float CandidateBoostedDoubleSecondaryVertexComputer::discriminator(const TagInfo
+    tau2_flightDistance2dSig= temp;
+
+    temp = tau1_vertexDeltaR;
+    tau1_vertexDeltaR= tau2_vertexDeltaR;
```
Is `tau2_vertexDeltaR = temp;` missing here? If not, the assignment `temp = tau1_vertexDeltaR;` is pointless and should be deleted.

ferencek replied:
`tau2_vertexDeltaR` is not used later so `tau2_vertexDeltaR = temp` would be a dead assignment and was removed in one of the previous commit after you made a comment about. You are now making a valid point that `temp = tau1_vertexDeltaR` is pointless and should be deleted. This is fixed now.

cvuosalo commented:
An extended test of workflow 134.805_RunMET2015C with 200 events against baseline CMSSW_8_0_0_pre1 shows no significant differences. There is a negligible change in the size of RECO event content, and Mini-AOD gets only a 0.3% increase in the size of patJets_slimmedJetsAK8, as expected.

ferencek committed:
removing dead assignment

cmsbuild commented:
Pull request #12442 was updated. @cvuosalo, @monttj, @cmsbuild, @slava77, @vadler, @davidlange6 can you please check and sign again.

cvuosalo commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cvuosalo commented:
@cmsbuild please test
Try again

cvuosalo commented:
@smuzaffar, @degano: Jenkins tests seem to be failing to complete on a consistent basis. This PR is one example.

smuzaffar commented:
Previous couple of tries failed due to issue communicating with github. anyway I have restarted the tests.

cmsbuild commented:
+1
Tested at: 4d93da778bbd0b26dcfcfc0506b3e38241bead04
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-12442/10060/summary.html

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-12442/10060/summary.html

cvuosalo commented:
+1

For #12442 4d93da778bbd0b26dcfcfc0506b3e38241bead04

Updating the boosted double SV tagger. #12445 and #12446 are the 76X and 74X versions of this PR, and both have already been approved.

The code changes are satisfactory, and Jenkins tests against baseline CMSSW_8_0_X_2015-12-01-2300 show no significant differences,

ferencek committed:
making the code thread-safe

Njettiness is not thread-safe and therefore should not be used as a data member
of an ESProducer

cmsbuild commented:
Pull request #12442 was updated. @cvuosalo, @monttj, @cmsbuild, @slava77, @vadler, @davidlange6 can you please check and sign again.

cvuosalo commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
-1
@ferencek This pull request cannot be automatically  merged, could you please rebase it?
 you can see the log for git cms-merge-topic here: 
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-12442/10160/git-merge-result

ferencek commented:
The changes in this PR were merged into 80X through a forward-port of #12445. Closing.
